[충남대 FE 김은수] 3단계 테마 상품 목록 페이지 구현하기

shushuburger opened the pull request:
안녕하세요! 충남대학교 FE 12팀 김은수입니다.

### 수행사항
- [x] 본인만의 기준으로 일관된 코드를 작성해주세요.
- [x] 기능 단위로 나누어 커밋을 해주세요.
- [x] 재사용성을 고려하여 컴포넌트, 비즈니스 로직을 구현해요.
- [x] 테마 상품 목록 페이지
  - [x] 선물하기 홈 > 선물 테마 섹션의 아이템을 클릭하면 테마 상품 목록 페이지로 연결돼요.
  - [x] **`/api/themes/:themeId/info`** API를 사용하여 선물 테마 섹션의 히어로 영역을 구현해요.
  - [x] 만약 테마 정보를 가져오는 API가 404 에러가 발생하면 선물하기 홈 페이지로 연결해요.
  - [x] **`/api/themes/:themeId/products`** API를 사용하여 상품 리스트를 구현해요.
  - [x] 무한 스크롤 기능을 구현해요.
  - [x] 상품 리스트가 없으면 빈 페이지를 보여줘요.
- [x] 2단계 피드백 반영
  - [x] isErrorWithMessage 유틸 함수로 분리
  - [x] api 관련 하드코딩 부분 상수화 처리
  - [x] useCategoryThemes.ts에서 useFetch() 라는 훅을 정의해서 사용
  - [x] CategoryContent 내부에서 useCategoryThemes를 호출
  - [x] OrderPage 내 로직 hook으로 분리
  - [x] ProductCard에서 호출 방식으로 ...product 활용
  - [x] ProductInfo에서 (product: ProductSummary) 방식 활용
  - [x] is는 변수에만 함수는 has를 사용하도록 변경
- [x] npm run build 성공적 수행 확인

shushuburger committed:
docs: 3단계 구현 요구사항 기록

shushuburger committed:
refactor(utils): isErrorWithMessage 함수 유틸 파일로 분리

- utils/error.ts에 공통 에러 유틸에 정리해주었습니다.
- LoginFormSection 내부 중복되는 함수를 제거했습니다.

shushuburger committed:
refactor(api): BASE_URL 기반 전체 API URL 상수 추가

- LOGIN_API_URL, ORDER_API_URL 등 전체 경로를 정의해주었습니다.
- getProductSummaryUrl 함수로 상품 요약 조회 경로를 만들어주었습니다.
- 하드코딩된 문자열 제거 후에 상수로 대체해주었습니다.

shushuburger committed:
refactor(hooks): useCategoryThemes 내부에 useFetch 함수 정의하여 fetch 로직 단순화

- useFetch를 통한 방식으로 hook이 사용되도록 수정해주었습니다.

shushuburger committed:
refactor(category): CategoryContent가 직접 useCategoryThemes 훅 호출하도록 변경

- themes props 제거로 불필요한 prop 전달을 최소화 시켰습니다.
- CategoryContent 내부에서 fetch 및 렌더링을 담당하도록 해 응집도를 향상시켰습니다.

shushuburger committed:
refactor(order): 상품 요약 fetch 로직 useProductSummary 훅으로 분리

- axios, useEffect 코드를 별도 훅으로 추출해주었습니다.

shushuburger committed:
refactor(product): ProductCard 호출부를 구조분해 방식에 맞게 수정

- product props 제거하고 ...product + rank 구조로 변경해주었습니다.
- ProductCard의 구조분해 props({ rank, ...product })와 일치하도록 수정해주었습니다.

shushuburger committed:
refactor(product): ProductInfo props 구조를 전체 객체 수용 방식으로 변경

- 구조분해 없이 product 객체 전체를 인자로 받도록 변경해주었습니다.
- OrderPage에서의 호출부는 ...product 방식으로 변경해주었습니다.

shushuburger committed:
refactor(naming): isErrorWithMessage → hasErrorMessage 함수명 변경

- 함수의 이름에 is대신 has를 사용하여 boolean 값과 혼동되는 것을 방지했습니다.
- 상태값으로 쓰이는 isExpanded는 그대로 유지했습니다.

shushuburger committed:
feat(route): 테마 상품 목록 페이지 라우트 추가

- /themes/:themeId 경로를 ROUTES 상수에 추가해주었습니다.
- ThemePage.tsx 기본 컴포넌트를 생성해두었습니다.

shushuburger committed:
feat(api): 테마 상세 정보 API 경로 및 URL 유틸 함수 추가

- getThemeInfoPath(themeId): /api/themes/:themeId/info 경로를 반환합니다.
- getThemeInfoUrl(themeId): BASE_URL 포함 전체 요청 URL을 생성해줍니다.

shushuburger committed:
feat(theme): 테마 상품 목록 페이지 초기 구현 완료

- /themes/:themeId 경로로 ThemePage 라우트를 수정해주었습니다.
- ThemeHero 컴포넌트에서 /api/themes/:themeId/info API 연동시켜주었습니다.
  - 404 응답 시 홈으로 navigate되도록 해주었습니다.
  - Banner를 활용해 title, description, 배경색을 지정해주었습니다.
- ThemeProductList 컴포넌트에서 /api/themes/:themeId/products API를 호출하도록 했습니다.
  - ProductCard, CardGrid를 활용해 상품 리스트 UI 구성했습니다.
  - 로딩, 에러, 빈 목록 상태에 따른 메시지를 각각 대응하도록 했습니다.
- getThemeInfoUrl, getThemeProductsUrl 유틸 함수로 API 경로 추출 상수화를 해주었습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
feat(category): 카테고리 클릭 시 해당 테마 상품 목록 페이지로 이동 기능 추가

- CategoryContent 컴포넌트에서 각 Item 클릭 시 /themes/:themeId 경로로 이동하도록 해주었습니다.
- useNavigate 훅 사용하여 라우팅 처리를 구현했습니다.
- README에 구현 완료한 내용을 표시했습니다.

shushuburger committed:
feat(theme): 테마별 상품 목록 무한 스크롤 기능 구현

- /api/themes/:themeId/products API를 cursor 기반으로 연동되도록 했습니다.
- 응답 구조에 맞춰 list, hasMoreList, cursor 값이 처리되도록 했습니다.
- IntersectionObserver를 이용한 무한 스크롤 적용을 해주었습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
fix(product): key 중복 경고 해결 및 랭킹 배지 표시 조건 추가

- ThemeProductList에서 중복 상품 제거 로직을 추가해주었습니다.
- ProductCard에서 rank 전달 시 key 충돌 방지를 위해 고유 id 유지하도록 했습니다.
- rank는 여전히 넘기되 hideRank로 시각적 표시 여부 제어 가능하도록 처리해주었습니다.

shushuburger committed:
style(theme): 테마 히어로 섹션 스타일 개선 및 설명 문구 추가

- themeInfo.name, title, description 순으로 시맨틱 구조를 정리해주었습니다.
- description이 있을 경우에만 조건부 렌더링이 일어나도록 했습니다.

shushuburger committed:
refactor(constants): 에러 메시지 및 정규표현식 상수화

- ThemeHero 컴포넌트 내 에러/로딩 텍스트를 validation.ts 상수로 추출해주었습니다.
- ThemeProductList 컴포넌트 내 상품 관련 에러/빈 데이터 텍스트를 상수로 대체했습니다.
- 기존의 loading 컴포넌트를 활용해 재사용성을 높였습니다.

shushuburger committed:
refactor(hook): theme 정보 조회 로직 useThemeInfo 훅으로 추출

- ThemeHero 컴포넌트 내 axios 호출, error/로딩 상태를 useThemeInfo로 분리해주었습니다.
- ThemeHero는 UI만 담당하는 컴포넌트로 역할을 분리해주었습니다.

shushuburger committed:
refactor(hook): 테마 상품 무한스크롤 로직 useThemeProducts 훅으로 추출

- ThemeProductList 컴포넌트의 상품 요청 및 observer 상태 관리를 useThemeProducts로 분리해주었습니다.
- UI는 데이터 렌더링만 담당하도록 해 책임 분리 및 재사용성을 높였습니다.

shushuburger committed:
refactor(type): Product, ProductSummary 타입을 공통 정의로 통합 활용

- useThemeProducts 및 ThemeProductList에서 중복된 Product 타입을 제거했습니다.
- theme과 관련된 타입도 차후 재사용성을 고려해 theme.ts로 추출해주었습니다.

shushuburger commented:
### 멘토님의 경험이 궁금합니다!!

이번에는 막 과제에 사용한 이론적인 내용과는 높은 관련이 있지 않은 궁금증입니다. 오타에 관한 궁금증인데요.

제가 3단계를 진행하면서 router 경로를 새로 등록을 해줄 때 themeId라고 적어야 하는데 themeid라고 적어버린 적이 있습니다.
themeId라는 단어를 꽤나 여러곳에서 사용을 했어서 '오타가 나지 않게 조심해야지'라고 생각을 했지만서도 
결국에는 themeid라는 오타를 딱 1개 발생시켰었는데요. routes.ts의 `THEME_PATH: '/themes/:themeId'`, 여기서요!

이것 때문에 올바른 curl 문으로 api 요청을 해도 npm run dev로 띄운 제 선물하기 사이트에서는 막연히 '불러오는 중...'이라는 문구만 떴었습니다.
postman 같은 곳에 요청을 하면 너무나도 성공적으로 잘 나오는데도 불구하고 
console.log에도 아무런 결과가 뜨지 않고 그렇다고 404 같이 api와의 소통에서 잘못된 점이 있다라는 error도 뜨지 않아서 
'아니!!! 대체 왜!!! 안 되는거야!!!'라는 막막함을 느꼈었습니다.

- 하드코딩을 없애자 -> 상수 처리

위 방법을 통해서 어느 정도 오타의 확률을 줄이긴 했지만 그럼에도 불구하고 발생할 수 있는게 오타라고 생각합니다.
변수명을 입력할 때 자동완성을 잘못 이용할 수도 있고 다양한 이유로 오타는 발생할 수 있으니깐요!

그리고 기존에 존재하는 prettier, eslint 같은 것들은 문자열 리터럴 내부의 오타는 잡아내지 못하는 걸로 알고 있습니다.
gpt한테도 물어보니 "타입 체크 대상이 아님, :나 /가 포함되면 단어로 인식하지 못할 수도 있음"과 같은 다양한 이유가 있었습니다.

문자열에 대해 발생하는 오타는 제가 더더 신경을 쓰는 것이 가장 현실적인 방법이라고 생각이 드는데요.

제가 이번에 시간을 이 부분에 막혀서 40분 정도를 소요하며 느낀 점은 
'문제가 발생했을 때 그 원인을 추론을 하는데 오타가 원인이라는 생각이 잘 들지 않아서 
배제하다보니 시간이 오래 걸린 것 아닐까'라는 생각입니다.

- 내가 로직을 잘못 짰나?
- 와이파이 연결이 안 좋나?
- ctrl+s를 안 눌렀나?

이 3개가 명확한 이유를 확인하기 어려운 경우 제가 가장 먼저 생각해내는 문제의 원인들입니다.

### 멘토님께서는 만약 코드 작업을 하다가 "명확한 error문이 없고 console에 data를 찍어보려하지만 찍히지 않을때"와 같이 바로 명확하게 '아! 이 부분이 문제구나!'라는 생각이 들기 어려운 경우에 어떤 식으로 접근하시는지 궁금합니다. "오타"는 몇 순위 정도로 떠올리시는지도 궁금하고요!

JunilHwang commented on README.md:
```diff
@@ -40,3 +40,27 @@
+  - [x] **`/api/themes/:themeId/products`** API를 사용하여 상품 리스트를 구현해요.
+  - [x] 무한 스크롤 기능을 구현해요.
+  - [x] 상품 리스트가 없으면 빈 페이지를 보여줘요.
+- [x] 2단계 피드백 반영
+  - [x] isErrorWithMessage 유틸 함수로 분리
+  - [x] api 관련 하드코딩 부분 상수화 처리
+  - [x] useCategoryThemes.ts에서 useFetch() 라는 훅을 정의해서 사용
+  - [x] CategoryContent 내부에서 useCategoryThemes를 호출
+  - [x] OrderPage 내 로직 hook으로 분리
+  - [x] ProductCard에서 호출 방식으로 ...product 활용
+  - [x] ProductInfo에서 (product: ProductSummary) 방식 활용
+  - [x] is는 변수에만 함수는 has를 사용하도록 변경
```
이런 부분도 정리해주시니까 좋네요 ㅎㅎ

shushuburger replied:
감사합니다!!

JunilHwang commented on src/utils/error.ts:
```diff
@@ -0,0 +1,17 @@
+export const hasErrorMessage = (
+  err: unknown
+): err is { response?: { data?: { message?: string } } } =>
+  typeof err === 'object' &&
+  err !== null &&
+  'response' in err &&
+  typeof (err as any).response?.data?.message === 'string';
```
아예 is 대신에 has를 쓰셨군요 ㅋㅋ
이것도 좋은 방법인 것 같아요!

shushuburger replied:
다행이네요! 감사합니다

JunilHwang commented on src/hooks/useProductSummary.ts:
```diff
@@ -0,0 +1,45 @@
+    fetchProduct();
+  }, [id, navigate]);
+
+  return { product, isLoading };
```
useCategoryThemes에서 만들어준 useFetch를 여기서 재활용할 수 있을것 같아요!

shushuburger replied:
useCategoryThemes에서 useFetch에 해당하는 부분을 useFetch.ts로 따로 분리해주고 export 해주었습니다!

JunilHwang commented on src/hooks/useThemeInfo.ts:
```diff
@@ -0,0 +1,35 @@
+    fetchThemeInfo();
+  }, [themeId, navigate]);
+
+  return { themeInfo, error };
```
훅으로 내보낼 때 "해체해서 사용할꺼야!" 라는 전제가 항상 깔려있는 것 처럼 느껴져요.
다만 이게 중복표현이 될 수 있다고 생각합니다.
가령,

```tsx
const { themeInfo } = useThemeInfo();
```
이런 모습으로 쓰이게 되는데요, themeInfo라는게 반복적으로 쓰이는 모습입니다.

그런데 기존에 추상화가 잘 되어있는 것들, 가령 html의 input을 참고해보면 이런 모습입니다.

<input value="">

이렇게 쓰이면 이상하겠죠?

`<input input="input">`

만약 themeInfo를 다루는 컴포넌트가 있다면 아마 이런식으로 선언될 것 같네요

<ThemeInfo themeInfo={themeInfo}>

이런 표현이 가독성을 높인다기보단 피로감을 느끼게 할 수 있으리라 생각해요!

이렇게 ThemeInfo의 value 혹은 data임을 보여주는 방식을 상상해보시면 좋답니다.

```jsx
const themeInfo = useThemeInfo;
<ThemeInfo data={themeInfo.data}>
<ThemeInfo {...themeInfo}>
```

shushuburger replied:
나타내고자 하는 것이 "테마의 정보에 관한 데이터"이다 보니 data를 사용하는 것이 제가 생각한 의미에 조금 더 적합하다고 생각했습니다!

```
export const useThemeInfo = (themeId: string | undefined) => {
  ...
  return { data, error }; // themeInfo -> data로 변경
};
```

현재는 useThemeInfo를 사용하는 ThemeHero.tsx에서 Section, TagText, Title과 같이 html 요소들에서 themeInfo라는 워딩을 사용하고 있지 않아. data에 themeInfo라는 별칭을 붙여 사용했습니다!

`  const { data: themeInfo, error } = useThemeInfo(themeId);`

차후 중복 표현이 되는 요소를 사용하게 된다면 별칭을 없애고 data로 사용하려고 합니다!

JunilHwang commented on src/hooks/useThemeProducts.ts:
```diff
@@ -0,0 +1,78 @@
+    return () => {
+      if (el) observer.unobserve(el);
+    };
+  }, [fetchProducts, hasMore, pending]);
```
이런 부분도 훅으로 분리해주면 좋을것 같아요!

shushuburger replied:
useIntersectionObserver.ts로 분리해주었습니다

JunilHwang commented on src/hooks/useThemeProducts.ts:
```diff
@@ -0,0 +1,78 @@
+
+      setProducts(prev => {
+        const merged = [...prev, ...newProducts];
+        const unique = Array.from(new Map(merged.map(p => [p.id, p])).values());
```
오호.. 이렇게 중복을 제거하는 방법도 좋네요 ㅎㅎ 덕분에 좋은 아이디어 얻어갑니다!!

JunilHwang commented on src/hooks/useThemeProducts.ts:
```diff
@@ -0,0 +1,78 @@
+  const [hasMore, setHasMore] = useState(true);
+  const [pending, setPending] = useState(false);
+  const [error, setError] = useState(false);
+  const observerRef = useRef<HTMLDivElement | null>(null);
```
이렇게 쓰이는 모습을 상상해보세요!
```jsx
const products = useFetch(fetchProducts)

useIntersectionObserver({ onIntersection: products.refetch });
);
```

JunilHwang commented on src/constants/api.ts:
```diff
@@ -6,3 +8,22 @@ export const CATEGORY_THEMES_API_PATH = '/api/themes';
+  `/api/themes/${themeId}/products`;
+
+export const getThemeProductsUrl = (themeId: string | number): string =>
+  `${BASE_URL}${getThemeProductsPath(themeId)}`;
```
API와 관련된 한경변수들을 constants 라는 파일에서 관리하는게 좋은 방법일까요?
hooks 폴더에 있는 hook에서 api를 호출하고 있고, 이 상수는 hooks 폴더에서만 쓰여요.
이렇게 넓은 스코프를 가질 필요가 없는거죠!

shushuburger replied:
hooks 내에 constants 폴더를 만든 다음 api.ts를 그 안으로 이동시켜주었습니다

JunilHwang commented on src/components/LoginFormSection.tsx:
```diff
@@ -47,7 +40,7 @@ const LoginFormSection = () => {
       await login({ email: email.value, password: password.value });
```
표현의 일관성이 지금 깨지는 모습이 보여요!

useLoginForm에서는 `isValid` 라는 표현을 사용하고,
email.validate()를 통해 쓰이는 값은 isOk 라는 표현을 사용합니다.

isValidEmail
isValidPassword

이런 모습이면 어떨까요?

shushuburger replied:
변경하였습니다!

JunilHwang commented on src/components/ThemeSection/ThemeProductList.tsx:
```diff
@@ -0,0 +1,57 @@
+      {pending && loading}
+      {hasMore && <ObserverTarget ref={observerRef} />}
+    </Wrapper>
+  );
```
이런 모습도 상상해볼 수 있답니다 ㅎㅎ

`if ~~ else if ~~`
이런 형태로 사용한느걸 "명령형 프로그래밍" 이라고 이야기한답니다.
그리고 보통 가독성을 헤치는 경우가 명령형 프로그래밍으로 작성되었을 때이고,
이걸 개선하는 방법이 "선언형 프로그래밍" 형태로 작성할 때라고 생각해요.

리액트는 선언형 프로그래밍 방식을 적극적으로 지원해줘요. 이렇게요!


```tsx
const products = useThemeProducts(themeId)
return (
    <WithApiUi
      data={products}
      loading={<EmptyText>{ERROR_MESSAGES.NO_PRODUCTS_AVAILABLE}</EmptyText>}
      error={<ErrorText>{ERROR_MESSAGES.FAILED_TO_LOAD_PRODUCTS}</ErrorText>}
    >
        <Wrapper>
            <CardGrid>
                {products.map((product, index) => (
                <ProductCard
                    key={product.id}
                    {...product}
                    rank={index + 1}
                    hideRank
                />
                ))}
            </CardGrid>
            {pending && loading}
            {hasMore && <ObserverTarget ref={observerRef} />}
        </Wrapper>
    </WithApiUi>
)
```

hooks와 UI를 결합해서 사용할 수도 있답니다 ㅎㅎ

```tsx
const products = useThemeProducts(themeId)
return (
    <products.WithApiUi
      data={products}
      loading={<EmptyText>{ERROR_MESSAGES.NO_PRODUCTS_AVAILABLE}</EmptyText>}
      error={<ErrorText>{ERROR_MESSAGES.FAILED_TO_LOAD_PRODUCTS}</ErrorText>}
    >
        <Wrapper>
            <CardGrid>
                {products.map((product, index) => (
                <ProductCard
                    key={product.id}
                    {...product}
                    rank={index + 1}
                    hideRank
                />
                ))}
            </CardGrid>
            {pending && loading}
            {hasMore && <ObserverTarget ref={observerRef} />}
        </Wrapper>
    </products.WithApiUi>
)
```

loading이나 Error의 조건이 데이터의 상태에 따라 달라질 수 있으니까요!

중요한건 "선언형으로 UI를 정의하는 방법" 이라고 생각해요.

shushuburger replied:
조건별 분기를 담당하는 WithApiUi 컴포넌트를 생성하였습니다.
이 컴포넌트는 data, error, loading, errorFallback 등의 props를 통해 상태에 따른 UI를 분기하며,
기존 ThemeProductList.tsx에서는 조건 처리 로직을 추상화하고 UI 구성에만 집중할 수 있도록 리팩토링하였습니다.

JunilHwang reviewed (changes requested):
안녕하세요 은수님!
마지막 3단계도 잘 진행해주셨네요 ㅎㅎ
다만 마지막 단계이고, 개선할 수 있는 부분 혹은 시도해볼 수 있는 부분이 많이 보여서 Request Changes로 남깁니다.

특히, 커스텀훅 만드는 연습을 지금 단계에서 많이 해주시면 좋겠어요.

> 멘토님께서는 만약 코드 작업을 하다가 "명확한 error문이 없고 console에 data를 찍어보려하지만 찍히지 않을때"와 같이 바로 명확하게 '아! 이 부분이 문제구나!'라는 생각이 들기 어려운 경우에 어떤 식으로 접근하시는지 궁금합니다. "오타"는 몇 순위 정도로 떠올리시는지도 궁금하고요!

오타는 스펠링과 관련된 플러그인을 적극적으로 이용해주세요. 저는 intellij(혹은 WebStorm)을 사용하는 편인데,

- 사용되지 않는 변수를 바로 알려주고
- 스펠링이 이상하면 알려주고

이런 기능이 내장되어 있답니다. 아마 vscode 에도 비슷한 기능이 분명 있으리라 생각해요 ㅎㅎ
사람이 직적접으로 해야 하는 일이 있고, 굳이 사람이 하지 않아도 되는 일이 분명 있어요.
오타를 찾아내는 부분도 사람이 굳이 하지 않아도 되는 일 중에 하나라고 생각합니다.

그리고 콘솔을 하나하나 사용하는 방법도 있지만,

console 대신 debbuger를 사용해보세요 ㅎㅎ
그냥 `debbuger`라고 코드 하단에 입력하면 된답니다.
그러면 디버거가 정의된 위치에서 코드의 실행이 멈추고 컨텍스트를 조회할 수 있어요.
어떤 값이 쓰이고 있는지 같은 것들!

혹은 지금 리액트로 작성되어있으니 [리액트 확장프로그램](https://chromewebstore.google.com/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=ko)을 적극적으로 사용해보시는것도 좋은 방법입니다!

shushuburger committed:
refactor(hooks): 공통 훅 적용으로 fetch 및 observer 로직 재사용

- useFetch 훅을 도입하여 useProductSummary와 useThemeInfo의 API 호출 로직을 변경해주었습니다.
- useIntersectionObserver 훅을 도입하여 useThemeProducts의 무한스크롤 관찰자 로직을 커스텀 훅으로 분리했습니다.

shushuburger committed:
refactor(hooks): useThemeProducts 내부에 useFetch 및 observer 통합 적용

- useFetch 훅을 내부에서 사용하여 데이터 요청 로직을 추상화해주었습니다.
- 외부에서는 단일 훅 사용 방식 유지하되 내부에서 필요한 부분을 다른 커스텀 훅 호출을 통해 해결하도록 했습니다.

shushuburger committed:
fix(order): useProductSummary에서 navigate 후 throw로 인한 무한 재렌더링 문제 해결

- fetch 실패 시 navigate 후 throw 대신 null을 반환하도록 해주었습니다.
- 404 응답 시 NOT_FOUND로 이동하도록 했습니다.
- useFetch 내부에서 반복 호출되던 문제를 제거해주었습니다.

shushuburger committed:
refactor(api): API URL 상수의 스코프를 hooks 전용으로 좁혀 위치 이동

- constants/api.ts를 hooks/constants/api.ts로 이동시켰습니다.

shushuburger committed:
refactor(login): 필드 유효성 검사 결과 변수명 일관성 있게 수정

- email.validate() 결과를 isEmailOk → isValidEmail로 변경해주었습니다.
- password도 동일하게 isValidPassword로 변경했습니다.
- useLoginForm 내부의 isValid와 표현 형식을 통일했습니다.

shushuburger committed:
refactor(ui): ThemeProductList를 선언형 UI 컴포넌트 WithApiUi로 감싸도록 변경

- error, empty 상태를 WithApiUi 컴포넌트에서 처리하도록 추상화해주었습니다.
- if 조건문 기반의 명령형 렌더링을 선언형으로 전환했습니다.

shushuburger committed:
fix(order): hook 내용 수정 후 null 문제 해결

shushuburger commented:
npm run build가 성공적으로 수행됨을 확인했습니다

shushuburger committed:
refactor(hook): useFetch 중복 내용 제거

JunilHwang commented on src/components/CategorySection/CategoryContent.tsx:
```diff
@@ -1,40 +1,37 @@
 
-  if (themes.error) {
+  if (error) {
     return <EmptyText>{ERROR_MESSAGES.FAILED_TO_LOAD_THEMES}</EmptyText>;
   }
 
-  if (!themes.data || themes.data.length === 0) {
+  if (!data || data.length === 0) {
     return <EmptyText>{ERROR_MESSAGES.NO_THEMES_AVAILABLE}</EmptyText>;
   }
```
여기에서도 WIthAPiUI 가 쓰일 수 있겠네요!

JunilHwang commented on src/components/ThemeSection/ThemeHero.tsx:
```diff
@@ -0,0 +1,54 @@
+  const { themeId } = useParams<{ themeId: string }>();
+  const { data: themeInfo, error } = useThemeInfo(themeId);
+
+  if (error)
+    return <ErrorText>{ERROR_MESSAGES.FAILED_TO_LOAD_THEMES}</ErrorText>;
+  if (!themeInfo) return loading;
+
+  return (
+    <Section style={{ backgroundColor: themeInfo.backgroundColor }}>
+      <TagText>{themeInfo.name}</TagText>
+      <Title>{themeInfo.title}</Title>
+      {themeInfo.description && (
+        <Description>{themeInfo.description}</Description>
+      )}
+    </Section>
+  );
```
여기서도!

JunilHwang commented on src/contexts/AuthContext.tsx:
```diff
@@ -43,10 +43,10 @@ export const AuthProvider = ({ children }: { children: ReactNode }) => {
+      password,
+    });
     const userData = res.data.data;
     sessionStorage.setItem(SESSION_STORAGE_KEY, JSON.stringify(userData));
```
스토리지 사용 방식도 나중에 개선해주시면 좋겠어요!

JunilHwang commented on src/hooks/useIntersectionObserver.ts:
```diff
@@ -0,0 +1,29 @@
+  enabled: boolean;
+}
+
+export const useIntersectionObserver = <T extends Element = HTMLElement>({
+  target,
+  onIntersect,
+  enabled,
+}: Props<T>) => {
+  useEffect(() => {
+    if (!enabled || !target.current) return;
+
+    const observer = new IntersectionObserver(
+      ([entry]) => {
+        if (entry.isIntersecting) {
+          onIntersect();
+        }
+      },
+      { threshold: 1.0 }
+    );
+
+    observer.observe(target.current);
+    return () => observer.disconnect();
+  }, [target, onIntersect, enabled]);
+};
```
disconnect 하는 조건도 추가해줄 수 있을 것 같아요!

JunilHwang replied:
가령, 모든 데이터를 불러왔을 때 disconnect를 해줘야하다보니..

아니면 onIntersect로 호출하는 함수가 내부적으로 처리할 수도 있겠네요 ㅎㅎ

JunilHwang commented on src/hooks/useProductSummary.ts:
```diff
@@ -0,0 +1,52 @@
+export const useProductSummary = (id: string | undefined) => {
+  const navigate = useNavigate();
+
+  const fetchProduct = useCallback(async (): Promise<
+    ProductSummary | undefined
+  > => {
+    if (!id) {
+      navigate(ROUTES.NOT_FOUND);
+      return undefined;
+    }
+
+    const res = await fetch(getProductSummaryUrl(id));
+
+    if (res.status === 404) {
+      navigate(ROUTES.NOT_FOUND);
+      return undefined;
+    }
+
+    if (!res.ok) {
+      toast.error(ERROR_MESSAGES.LOAD_PRODUCT_FAIL, {
+        toastId: 'product-load-fail',
+      });
+      navigate(ROUTES.HOME);
+      return undefined;
+    }
+
+    const json = await res.json();
+    return json.data as ProductSummary;
+  }, [id, navigate]);
+
+  const { data: product, pending: isLoading } = useFetch<
+    ProductSummary | undefined
+  >(fetchProduct);
```
fetchProduct가 많은 일을 하고 있어요!
useFetch에게 `onError` 같은 옵션을 하나 넘겨주면 어떨까 싶네요 ㅎㅎ

JunilHwang commented on src/hooks/useThemeInfo.ts:
```diff
@@ -0,0 +1,30 @@
+  const fetchThemeInfo = useCallback(async () => {
+    if (!themeId) throw new Error('themeId is undefined');
+
+    const res = await fetch(getThemeInfoUrl(themeId));
+    if (res.status === 404) {
+      navigate(ROUTES.NOT_FOUND);
+      throw new Error('404 Not Found');
+    }
+    if (!res.ok) {
+      throw new Error('Theme info fetch failed');
+    }
```
잘 보면 반복적으로 쓰이는 패턴이 있어요!

JunilHwang commented on src/hooks/useThemeInfo.ts:
```diff
@@ -0,0 +1,30 @@
+export const useThemeInfo = (themeId: string | undefined) => {
+  const navigate = useNavigate();
+
+  const fetchThemeInfo = useCallback(async () => {
+    if (!themeId) throw new Error('themeId is undefined');
+
+    const res = await fetch(getThemeInfoUrl(themeId));
+    if (res.status === 404) {
+      navigate(ROUTES.NOT_FOUND);
+      throw new Error('404 Not Found');
+    }
+    if (!res.ok) {
+      throw new Error('Theme info fetch failed');
+    }
+
+    const json = await res.json();
+    return json.data as ThemeInfo;
+  }, [themeId, navigate]);
```
```suggestion
  const fetchThemeInfo = useApi(themeService.fetchTemeInfo, { onNotFound: () => ..., onError: () => ... });
```

내부적으로 notfound를 처리하는 기본 로직을 만들어준 다음, onNotFound를 추가로 호출하는 방식으로 추상화할 수 있답니다 ㅎㅎ

JunilHwang commented on src/hooks/useThemeProducts.ts:
```diff
@@ -0,0 +1,59 @@
+const LIMIT = 10;
+
+export const useThemeProducts = (themeId: string | undefined) => {
+  const [products, setProducts] = useState<Product[]>([]);
+  const [cursor, setCursor] = useState(0);
+  const [hasMore, setHasMore] = useState(true);
+  const observerRef = useRef<HTMLDivElement | null>(null);
+
+  const fetchProducts = useCallback(async () => {
+    if (!themeId) throw new Error('themeId is undefined');
+
+    const res = await fetch(
+      `${getThemeProductsUrl(themeId)}?cursor=${cursor}&limit=${LIMIT}`
+    );
+    if (!res.ok) throw new Error('Failed to fetch theme products');
+
+    const json = await res.json();
+    return json.data as {
+      list: Product[];
+      hasMoreList: boolean;
+      cursor: number;
+    };
+  }, [themeId, cursor]);
+
+  const { data, pending, error, refetch } = useFetch(fetchProducts);
+
+  useEffect(() => {
+    if (!data) return;
+    const { list, hasMoreList, cursor: nextCursor } = data;
+
+    setProducts(prev => {
+      const merged = [...prev, ...list];
+      const unique = Array.from(new Map(merged.map(p => [p.id, p])).values());
+      return unique;
+    });
+    setHasMore(hasMoreList);
+    setCursor(nextCursor);
+  }, [data]);
```
useFetch와 엮어서 사용하는게 어려웠나보군요 ㅎㅎ
```jsx
const { call } = useFetch(fn, { auto: false })
```

이런걸 하나 만들어서 사용해주면 어떨까 싶기도 해요!

JunilHwang reviewed (approved):
안녕하세요 은수님!
피드백을 굉장히 빠르게 반영해주셨네요 ㅋㅋ 고생하셨습니다!!
WithApiUI를 활용할 수 있는 부분이 많은 것 같은데, 한 군데서만 사용해서 아쉬워요.
이 외에도 천천히 살펴보면 추상화할 수 잇는 훅이 많은데, 제가 남긴 피드백에 대해서만 반영한것도 조금 아쉬운 부분이네요..!
어떤 부분을 분리하면 좋을지 은수님께서 스스로 연습하는 과정이 필요해보여요.

이번 미션은 여기서 마무리하도록 하겠습니다!
고생하셨어요 은수님~!
